### PR TITLE
executor: implements nested loop join.

### PR DESCRIPTION
**Pull request by hanfei1991**
I implement nested loop join that would be used by apply.
@shenli @coocood @zimulala @winoros @lamxTyler PTAL

### CODE REVIEW

File: executor/executor_join.go [hunk 1]
```diff
@@ -25,6 +25,12 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
+var (
+	_ Executor = &HashAggExec{}
```
Comment (shenli) [inline executor/executor_join.go]: Why not put it in executor_agg.go?

File: executor/executor_join.go [hunk 2]
```diff
@@ -382,6 +388,148 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+	fetchBigRow() (*Row, error)
+	// prepare reads all records from small Exec and stores them.
+	prepare() error
+	// doJoin fetch a row from big exec and get all the rows matches the on condition.
+	doJoin(*Row) ([]*Row, error)
+}
+
+// NestedLoopJoinExec implements nested-loop algorithm for join.
+// TODO: Currently we only implements inner join for it. We will complete it in future.
+type NestedLoopJoinExec struct {
```
Comment (shenli) [inline executor/executor_join.go]: When will we use NestedLoopJoinExec instead of HashJoin?

Comment (hanfei1991) [inline executor/executor_join.go]: When we use apply, it may contain this but hash join

File: executor/executor_join.go [hunk 3]
```diff
@@ -382,6 +388,148 @@ func (e *HashJoinExec) Next() (*Row, error) {
 	return row, nil
 }
 
+// joinExec is the common interface of join algorithm except for hash join.
+type joinExec interface {
+	Executor
+
+	// fetchBigRow fetches a valid big row from big Exec.
```
Comment (shenli) [inline executor/executor_join.go]: a valid big row -> a valid row

File: executor/executor_join.go [hunk 4]
```diff
@@ -382,6 +388,148 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+		if e.cursor < len(e.resultRows) {
+			retRow := e.resultRows[e.cursor]
+			e.cursor++
+			return retRow, nil
+		}
+		bigRow, err := e.fetchBigRow()
+		if bigRow == nil || err != nil {
+			return bigRow, errors.Trace(err)
+		}
+		e.resultRows, err = e.doJoin(bigRow)
```
Comment (shenli) [inline executor/executor_join.go]: Check the error.

File: executor/executor_join.go [hunk 5]
```diff
@@ -521,32 +662,97 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
...
+
+	for {
+		bigRow, err := e.fetchBigRow()
+		if bigRow == nil || err != nil {
+			return bigRow, errors.Trace(err)
 		}
-		if matched {
-			return bigRow, nil
+		resultRows, err := e.doJoin(bigRow)
+		if len(resultRows) > 0 {
```
Comment (shenli) [inline executor/executor_join.go]: Check error first.

File: executor/executor_join.go [hunk 6]
```diff
@@ -521,32 +666,109 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
...
+	// For the non-auxMode subquery, we return the matching row only.
+	if e.auxMode {
+		if isNull {
+			bigRow.Data = append(bigRow.Data, types.NewDatum(nil))
+		} else {
+			bigRow.Data = append(bigRow.Data, types.NewDatum(matched))
 		}
-		if e.anti && !isNull {
-			matched = !matched
+		return []*Row{bigRow}, nil
```
Comment (shenli) [inline executor/executor_join.go]: Can we add an interface named doSemiJoin in joinExec? So we do not have to create a slice here.

File: executor/executor_join.go [hunk 7]
```diff
@@ -521,32 +666,109 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
...
+			return row, nil
+		}
+		bigRow, err := e.join.fetchBigRow()
+		if bigRow == nil || err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, col := range e.outerSchema {
+			*col.Data = bigRow.Data[col.Index]
+		}
+		e.innerExec.Close()
```
Comment (winoros) [inline executor/executor_join.go]: Where is the e.innerExec used? Is it the smallExec of e.join?

File: executor/executor_join.go [hunk 8]
```diff
@@ -521,32 +666,109 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
...
+}
+
+// ApplyJoinExec is the new logic of apply.
+// TODO: I will add test for it after refactoring the plan part of apply.
+type ApplyJoinExec struct {
+	join        joinExec
+	innerExec   Executor
+	outerSchema []*expression.CorrelatedColumn
+	cursor      int
+	resultRow   []*Row
```
Comment (winoros) [inline executor/executor_join.go]: resultRow -> resultRows?

File: executor/executor_join.go (hunk 7 above)
Comment (hanfei1991) [inline executor/executor_join.go]: Yeah, innerExec equals to smallExec. In Semi-Join it's called "inner".

**Comment by hanfei1991** [comment]
@shenli @winoros  PTAL

File: executor/executor_join.go [hunk 9]
```diff
@@ -382,6 +389,151 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+}
+
+func (e *NestedLoopJoinExec) fetchBigRow() (*Row, error) {
+	for {
+		bigRow, err := e.BigExec.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if bigRow == nil {
+			e.BigExec.Close()
```
Comment (lamxTyler) [inline executor/executor_join.go]: Should we check error here?

File: executor/executor_join.go [hunk 10]
```diff
@@ -382,6 +389,151 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+
+		matched := true
+		if e.BigFilter != nil {
+			matched, err = expression.EvalBool(e.BigFilter, bigRow.Data, e.Ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if matched {
+			return bigRow, err
```
Comment (lamxTyler) [inline executor/executor_join.go]: Is `return bigRow, nil` better? The `err` here makes me confused.

File: executor/executor_join.go [hunk 11]
```diff
@@ -382,6 +389,151 @@ func (e *HashJoinExec) Next() (*Row, error) {
...
+// Prepare runs the first time when 'Next' is called and it reads all data from the small table and stores
+// them in a slice.
+func (e *NestedLoopJoinExec) prepare() error {
+	for {
+		row, err := e.SmallExec.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			e.SmallExec.Close()
```
Comment (lamxTyler) [inline executor/executor_join.go]: check error?

File: executor/executor_join.go [hunk 12]
```diff
@@ -521,32 +669,110 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
...
+			return row, nil
+		}
+		bigRow, err := e.join.fetchBigRow()
+		if bigRow == nil || err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, col := range e.outerSchema {
+			*col.Data = bigRow.Data[col.Index]
+		}
+		e.innerExec.Close()
```
Comment (lamxTyler) [inline executor/executor_join.go]: check error?

**Comment by shenli** [comment]
@lamxTyler PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**